### PR TITLE
Fabio looker/marketo endpoint enhancements

### PR DESCRIPTION
**Pull request by fabio-looker**
PR adds additional modes to the Marketo action, specifically:
- Being able to update leads WITHOUT providing a campaign ID
- Being able to update leads and add them to a list
- Being able to update leads and remove them from a list

image

### CODE REVIEW

**Comment by fabio-looker** [comment]
Fixing TS lint stuff...

**Comment by dwmintz** [comment]
cc: @phillipperalez

File: src/actions/marketo/marketo_transaction.ts [hunk 1]
```diff
@@ -10,21 +10,58 @@ interface Result {
...
+    const subactionIds =
+      (request.formParams.subactionIds !== undefined ? request.formParams.subactionIds : "")
+      .split(/\s*,\s*/)
+      .filter(Boolean)
+    switch (request.formParams.subaction) {
+      case undefined:
+        // The older version of the action assumed an "addToCampaign" subaction
+        this.campaignIds = [request.formParams.campaignId !== undefined ? request.formParams.campaignId : ""]
+          .filter(Boolean) // Note singular parameter name
+        break
```
Comment (dwmintz) [inline src/actions/marketo/marketo_transaction.ts]: @phillipperalez this logic is critical to make sure that existing schedules that were created with the existing version of the Marketo action do not break.  It defaults to the update record and add to campaign action, which is the only one that the current version supports.

**Comment by apstrong** [comment]
@jdavidson or @wilg could one of you review this? Phillip's focusing on alerts work this week

**Comment by jdavidson** [comment]
I'll try to take a look in a bit. Why not make each of the four different behaviors 4 different actions?

**Comment by fabio-looker** [comment]
@jdavidson No particular reason. It just seemed like a pretty big code change to make separate actions for each.

**Comment by dwmintz** [comment]
I strongly prefer having them all in one from a UI perspective. Just feels like it'd be overwhelming in the picker to start having many of each logo

Best,
-d

**Comment by fabio-looker** [comment]
Please hold on review until tomorrow. I need to fix bug that I introduced during my "fixing" of Typescript linter warnings. And also the test that I thought should catch this case should be expanded...

**Comment by fabio-looker** [comment]
Btw, this should now be good for review

**Comment by dwmintz** [comment]
Just pinging @wilg and @jdavidson on this to see if one of you can take a look.

**Comment by maxreid1** [comment]
I have a prospect in trial very interested in this - do we have an ETA on it?

**Comment by dwmintz** [comment]
Approving the readme changes.

**Comment by dwmintz** [comment]
@fabio-looker we did some testing of this on dev, and while generally it looks great, there's one bug that I think we need to fix before releasing. When we upgraded from v1 to v2, we used a test schedule to make sure that it didn't break, and unfortunately it did :-(

The behavior I saw was that upon upgrade, the existing schedule correctly kept the custom lookup field that I'd entered for the saved schedule, but it lost the campaign_id (it went to blank in v2) and the dropdown defaulted to `none`, rather than `addCampaign` (which it will always be if it's coming from v1).

Is that something you can take a look at?

**Comment by fabio-looker** [comment]
Certainly. I'm at the hackathon today. If I take a look by Friday is that alright?

**Comment by dwmintz** [comment]
Yup! Thanks.

Best,
-d

**Pull request merged**